### PR TITLE
Refactor of CaseReader **Backwards incompatible**

### PR DESCRIPTION
**Pull request by swryan**
Pivotal #160117632

Also: rewrite case_recording docs (Pivotal #160117673)

### CODE REVIEW

**Comment by kmarsteller** [comment]
Tests finally passing, just need one more review.

File: openmdao/recorders/sqlite_recorder.py [hunk 1]
```diff
@@ -26,18 +25,20 @@
 
 
 """
-SQL case output format version history.
----------------------------------------
+SQL case database version history.
+----------------------------------
+5 -- OpenMDAO 2.5
```
Comment (Kenneth-T-Moore) [inline openmdao/recorders/sqlite_recorder.py]: There is so much here that I can't really tell, but did you add a test that verifies we can read a version 4 recording?

Comment (swryan) [reply openmdao/recorders/sqlite_recorder.py]: Good point... I did not.   Will cook up a new test.

Comment (swryan) [reply openmdao/recorders/sqlite_recorder.py]: done

**Pull request merged**